File: clients/kotlin/src/main/kotlin/streamforge/ZSet.kt
package streamforge

import com.google.gson.Gson

/** One live row: an arbitrary bag of columns. Stays a Map -- there is no pandas equivalent here,
 * and typing it would fight the fact that ad-hoc tables (`adhoc_*`) exist created at runtime. */
typealias Row = Map<String, Any?>

/** One Z-set delta: a row entering (+1) or leaving (-1) a table's output. Weight is int64 on the
 * wire (TableDelta.weight / TableRow.weight in streamforge.proto). */
data class RowDelta(val row: Row, val weight: Long)

/** One batch of deltas as either transport frames it: `TableDeltaBatch` on gRPC,
 * `tableDelta(name, deltas, seq)` on SignalR. `seq` is a per-subscription batch counter, not
 * comparable across transports or to a snapshot's own `seq` -- see [ZSet]'s docstring. */
data class DeltaBatch(val deltas: List<RowDelta>, val seq: Long)

private val canonicalGson = Gson()

/** Stable identity for a row's exact content: sorted (key, value) pairs, JSON-serialized. Two
 * rows are the "same" delta target iff this string matches -- not any single column. */
fun canonicalKey(row: Row): String = canonicalGson.toJson(row.toSortedMap(compareBy { it }))

/**
 * The row's logical-identity ("group") key, or null when supersession does not apply.
 *
 * - `keyFields` non-null and non-empty: joins each field's value. Two canonical rows sharing the
 *   same group key are the same logical entity at different times (an updated MTM tick, a
 *   LATEST BY row) -- asserting a new one deletes the group's previous row even though nothing on
 *   the wire explicitly retracted it.
 * - `keyFields == emptyList()`: a global aggregate -- exactly one row, one constant group `"*"`.
 * - `keyFields == null`: the table's key is unknown and none was given. Deliberately NOT "guess
 *   the first column" (the browser console's fallback, wrong for a composite key or a global
 *   aggregate) -- this falls back to whole-row identity instead. Never guess a key.
 */
fun groupKeyOf(row: Row, keyFields: List<String>?): String? {
    if (keyFields == null) return null
    if (keyFields.isEmpty()) return "*"
    return keyFields.joinToString("|") { f ->
        if (row.containsKey(f)) "$f=${canonicalGson.toJson(row[f])}" else "$f=undefined"
    }
}

/**
 * Live reduced state for one table: canonicalKey -> (row, weight), plus a groupKey ->
 * canonicalKey index for supersession. Ported literally from `lib/streamforge/live-table.ts`
 * (via the Python client's `_zset.py`, whose module docstring is the fullest account of why this
 * looks the way it does) -- pure state, no I/O, so it is testable with handcrafted fixtures alone
 * (see the conformance suite in `../conformance/zset-cases.json`).
 *
 * Two hazards this defends against:
 *
 * 1. Weight is SUMMED per canonical identity across every delta seen; a summed weight <= 0
 *    removes the row outright rather than going negative. This is what makes retract-then-assert
 *    and assert-then-retract (arrival order is not guaranteed) both converge to the same state.
 * 2. Subscribe races the initial snapshot. The caller ([LiveTable]) buffers everything until the
 *    snapshot is in hand, seeds from it (dropping weight<=0 rows and resolving any supersession
 *    the snapshot itself straddled), then replays buffered batches -- except ones the snapshot
 *    already reflects. There is no shared sequence counter between a snapshot read and the delta
 *    stream (measured on different scales), so [alreadyReflected] is a CONTENT heuristic, not a
 *    seq comparison: a buffered batch is skipped only when EVERY one of its retractions targets a
 *    row the snapshot does not contain.
 */
class ZSet(private val keyFields: List<String>?) {
    private val map = LinkedHashMap<String, Pair<Row, Long>>() // canonicalKey -> (row, weight)
    private val groupIndex = HashMap<String, String>() // groupKey -> canonicalKey

    /** Current live rows. Collapses to one row per group when `keyFields` is known -- apply()/
     * seed() already maintain the one-canonical-key-per-group invariant, but a consumer must
     * never see a group surface twice regardless of that. */
    fun rows(): List<Row> {
        if (keyFields == null) return map.values.map { it.first }
        val byGroup = LinkedHashMap<String, Row>()
        for ((_, entry) in map) {
            val (row, _) = entry
            val gk = groupKeyOf(row, keyFields) ?: canonicalKey(row)
            byGroup[gk] = row
        }
        return byGroup.values.toList()
    }

    /** Apply one batch of deltas in place. Returns the canonical keys newly asserted (weight
     * summed > 0 after this batch) -- for on_change/flash tracking. */
    fun apply(deltas: List<RowDelta>): Set<String> {
        val touched = LinkedHashSet<String>()
        for ((row, weight) in deltas) {
            val key = canonicalKey(row)
            val groupKey = groupKeyOf(row, keyFields)
            val prevWeight = map[key]?.second ?: 0L
            val nextWeight = prevWeight + weight
            if (nextWeight <= 0) {
                map.remove(key)
                if (groupKey != null && groupIndex[groupKey] == key) groupIndex.remove(groupKey)
            } else {
                if (groupKey != null) {
                    val staleKey = groupIndex[groupKey]
                    if (staleKey != null && staleKey != key) map.remove(staleKey)
                    groupIndex[groupKey] = key
                }
                map[key] = row to nextWeight
                touched.add(key)
            }
        }
        return touched
    }

    /** Reset state and seed from a snapshot read (GET /rows or TableService.Rows). Mirrors
     * apply()'s rules: a weight<=0 row is not part of the snapshot at all, and a group keeps only
     * its newest row -- a snapshot read mid-update can carry both sides of a supersession. */
    fun seed(snapshotRows: List<RowDelta>) {
        map.clear()
        groupIndex.clear()
        for ((row, weight) in snapshotRows) {
            if (weight <= 0) continue
            val key = canonicalKey(row)
            val groupKey = groupKeyOf(row, keyFields)
            if (groupKey != null) {
                val staleKey = groupIndex[groupKey]
                if (staleKey != null && staleKey != key) map.remove(staleKey)
                groupIndex[groupKey] = key
            }
            map[key] = row to weight
        }
    }

    /** True when this buffered batch's effect is already visible in the (just-seeded) current
     * state. A batch with no retractions is never considered reflected -- an assert-only batch is
     * always safe, and possibly necessary, to replay. */
    fun alreadyReflected(deltas: List<RowDelta>): Boolean {
        val retractions = deltas.filter { it.weight < 0 }
        if (retractions.isEmpty()) return false
        for ((row, _) in retractions) {
            val weight = map[canonicalKey(row)]?.second ?: 0L
            if (weight > 0) return false
        }
        return true
    }
}

File: clients/python/src/streamforge/_zset.py
"""The Z-set reducer -- ported literally from lib/streamforge/live-table.ts (otc-terms), itself a
port of crates-foundation's own web/src/hooks/useTableRows.ts. Pure functions/state, no I/O:
this module never touches a socket, so it is testable with handcrafted fixtures alone.

Hazards, and why the code below looks the way it does:

- A row's identity for retract/assert matching is the WHOLE row's content, not any one column:
  `canonical_key` is a stable serialization of every (key, value) pair, sorted by key. Weight is
  SUMMED per canonical identity across every delta seen; a summed weight <= 0 removes the row
  outright rather than going negative. This is what makes retract-then-assert and
  assert-then-retract (arrival order is not guaranteed) both converge to the same state.

- A logical key ("group", `key_fields`) can SUPERSEDE: two different canonical rows sharing the
  same key_fields values are the same logical entity at different times (an updated MTM tick, a
  LATEST BY row). When a new row for a group is asserted, the group's PREVIOUS canonical row is
  deleted even though its own weight was never explicitly retracted on the wire -- the retraction
  is implied by the new assert superseding it. `group_key_of` returns `"*"` for `key_fields=[]`
  (a global aggregate: one row, one group), and `None` for `key_fields=None` (unknown table, no
  override given) -- see its own docstring for why that case does NOT fall back to "first column"
  the way the browser console does.

- Subscribe races the initial snapshot: deltas can arrive before, during or after the snapshot
  read lands. The caller (live.py) buffers everything until the snapshot is in hand, seeds state
  from it (dropping weight<=0 rows and resolving any supersession the snapshot itself straddled --
  a snapshot read mid-update can carry both the old and new row of a group), and then replays the
  buffered batches -- except ones the snapshot has ALREADY reflected. There is no shared sequence
  counter between the snapshot read (`/rows`' `seq`) and the delta stream (`TableDeltaBatch.seq`
  is a per-subscription batch counter on a completely different scale -- measured ~860 vs ~15,000
  at the same instant in useTableRows.ts) so "already reflected" cannot be seq-based. Instead:
  `already_reflected` is a CONTENT heuristic -- a buffered batch is skipped only when EVERY one of
  its retractions targets a row the snapshot does not contain (i.e. the snapshot already dropped
  it). Replaying it anyway would double-apply a retraction the snapshot already reflects, which
  for a plain reduce-by-weight is harmless, but for a LATEST BY group it can delete the WRONG
  (newer) row out from under the group index. wishlist #20 (a shared epoch on both the snapshot
  and the delta stream) is the fix that would make this exact instead of a heuristic.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

Row = dict[str, Any]
Delta = tuple[Row, int]  # (row, weight)

_MISSING = object()


def canonical_key(row: Row) -> str:
    """Stable identity for a row's exact content: sorted (key, value) pairs, JSON-serialized."""
    keys = sorted(row.keys())
    return json.dumps([[k, row[k]] for k in keys], separators=(",", ":"), default=str)


def group_key_of(row: Row, key_fields: list[str] | None) -> str | None:
    """The row's logical-identity ("group") key, or None when supersession does not apply.

    - `key_fields` is a real (possibly empty) list: `[]` is a global aggregate -- exactly one row,
      one constant group `"*"`. A non-empty list joins each field's value.
    - `key_fields is None`: the table's key is unknown and none was given. Deliberately NOT the
      browser console's fallback (live-table.ts's `groupKeyOf` uses the row's first column, which
      is right for a LATEST BY table that leads with its key and silently wrong for a composite
      key or a global aggregate). This client instead falls back to whole-row identity -- no
      column is ever guessed, and simply never superseding is the safe failure mode (design doc
      "Key fields" section: "never guess the first column"). Returning None here means
      apply()/seed() skip all group bookkeeping for the row, so canonical_key alone is its
      identity, which is exactly whole-row identity.
    """
    if key_fields is None:
        return None
    if len(key_fields) == 0:
        return "*"
    parts = []
    for f in key_fields:
        value = row[f] if f in row else _MISSING
        parts.append(f"{f}={json.dumps(value, default=str) if value is not _MISSING else 'undefined'}")
    return "|".join(parts)


@dataclass
class ZSet:
    """Live reduced state for one table: canonical_key -> (row, weight), plus a group_key ->
    canonical_key index for supersession. `key_fields=None` disables supersession entirely
    (whole-row identity); `key_fields=[]` is a global aggregate."""

    key_fields: list[str] | None
    _map: dict[str, Delta] = field(default_factory=dict)
    _group_index: dict[str, str] = field(default_factory=dict)

    def rows(self) -> list[Row]:
        """Current live rows. Collapses to one row per group when key_fields is known -- a
        defensive step mirroring live-table.ts's flushToState: apply()/seed() already maintain
        the one-canonical-key-per-group invariant, but a consumer must never see a group surface
        twice regardless."""
        if self.key_fields is None:
            return [row for row, _weight in self._map.values()]
        by_group: dict[str, Row] = {}
        for row, _weight in self._map.values():
            gk = group_key_of(row, self.key_fields) or canonical_key(row)
            by_group[gk] = row
        return list(by_group.values())

    def apply(self, deltas: list[Delta]) -> list[str]:
        """Apply one batch of (row, weight) deltas in place. Returns the canonical keys that were
        newly asserted (weight summed > 0 after this batch) -- for on_change/flash tracking."""
        touched: list[str] = []
        for row, weight in deltas:
            key = canonical_key(row)
            group_key = group_key_of(row, self.key_fields)
            _prev_row, prev_weight = self._map.get(key, (row, 0))
            next_weight = prev_weight + weight
            if next_weight <= 0:
                self._map.pop(key, None)
                if group_key is not None and self._group_index.get(group_key) == key:
                    del self._group_index[group_key]
            else:
                if group_key is not None:
                    stale_key = self._group_index.get(group_key)
                    if stale_key is not None and stale_key != key:
                        self._map.pop(stale_key, None)
                    self._group_index[group_key] = key
                self._map[key] = (row, next_weight)
                touched.append(key)
        return touched

    def seed(self, snapshot_rows: list[Delta]) -> None:
        """Reset state and seed from a snapshot read (GET /rows). Mirrors apply()'s rules: a
        weight<=0 row is not part of the snapshot at all, and a group keeps only its newest row --
        a snapshot read mid-update can carry both sides of a supersession."""
        self._map = {}
        self._group_index = {}
        for row, weight in snapshot_rows:
            if weight <= 0:
                continue
            key = canonical_key(row)
            group_key = group_key_of(row, self.key_fields)
            if group_key is not None:
                stale_key = self._group_index.get(group_key)
                if stale_key is not None and stale_key != key:
                    self._map.pop(stale_key, None)
                self._group_index[group_key] = key
            self._map[key] = (row, weight)

    def already_reflected(self, deltas: list[Delta]) -> bool:
        """True when this buffered batch's effect is already visible in the (just-seeded)
        current state -- see the module docstring's "Subscribe races the initial snapshot"
        paragraph. A batch with no retractions is never considered reflected (an assert-only
        batch is always safe, and possibly necessary, to replay)."""
        retractions = [(row, w) for row, w in deltas if w < 0]
        if not retractions:
            return False
        for row, _weight in retractions:
            key = canonical_key(row)
            _row, weight = self._map.get(key, (row, 0))
            if weight > 0:
                return False
        return True
